### PR TITLE
Display correct time format on availability page

### LINKED ISSUE #2082: Time format not respected
Found a bug? Please fill out the sections below. 👍

### Issue Summary

In my profile I've selected "12 Hour" time format, by on my events it lists available slots in 24 Hour time format (13:00, etc.). I could not find a time format setting on the event setup itself, which is what led me to believe the profile setting controlled this. Perhaps there is another location I have missed?

### Steps to Reproduce

1. Set time format to 12 Hour in profile settings
2. Expected to see time slots available for events in 12 Hour format, but did not

Any other relevant information. For example, why do you consider this a bug and what did you expect to happen instead?

### Technical details

- In case it is relevant, I experienced this issue on the hosted "app.cal.com" version (free saas version)

### PR DESCRIPTION
**Pull request by joeauyeung**
## What does this PR do?

This PR displays the proper time format on the availability page.

Fixes #2082

**Environment**: Staging(main branch) / Production

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

- Go to create a booking and choose a date
- The times displayed should be in 12h or 24h depending on your location
- In `timeFormat.ts` and comment out line 11.
- You can now change `locale` to "en-US" for 12h time or "en-GB" for 24h time

## Checklist

- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by zomars — CHANGES_REQUESTED** [review]
Leaving a video review, this PR is not addressing the expected behavior.

We should prioritize the user settings over the browser defaults

https://www.loom.com/share/4a3d3f8eb52f456882cc55e59a98521e

**Comment by joeauyeung** [comment]
> Leaving a video review, this PR is not addressing the expected behavior.
> 
> We should prioritize the user settings over the browser defaults
> 
> [loom.com/share/4a3d3f8eb52f456882cc55e59a98521e](https://www.loom.com/share/4a3d3f8eb52f456882cc55e59a98521e)

Thanks for the video, helps a lot for context. Just to confirm when displaying available times we should see if there are user settings available and use that time format option. If no user found, then use the browser locale to set the time format.

**Comment by zomars** [comment]
> > Leaving a video review, this PR is not addressing the expected behavior.
> > We should prioritize the user settings over the browser defaults
> > [loom.com/share/4a3d3f8eb52f456882cc55e59a98521e](https://www.loom.com/share/4a3d3f8eb52f456882cc55e59a98521e)
> 
> Thanks for the video, helps a lot for context. Just to confirm when displaying available times we should see if there are user settings available and use that time format option. If no user found, then use the browser locale to set the time format.

Yes, we could take this a step further and add a "Use browser default settings" as an option, similarly on how we use it for our dark/light mode dropdown.

File: apps/web/lib/timeFormat.ts [hunk 1]
```diff
@@ -4,9 +4,27 @@
...
+import { localStorage } from "@calcom/lib/webstorage";
+
 export const isBrowserLocale24h = () => {
+  const localStorageTimeFormat = localStorage.getItem("timeOption.is24hClock");
+  // If time format is already stored in the browser then retrieve and return early
+  if (localStorageTimeFormat === "true") {
+    return true;
+  } else if (localStorageTimeFormat === "false") {
+    return false;
+  }
```
Comment (joeauyeung) [inline apps/web/lib/timeFormat.ts:16]: Grabbing from local storage is now used in all areas this function is called

File: apps/web/lib/timeFormat.ts [hunk 2]
```diff
@@ -4,9 +4,27 @@
...
-  return !new Intl.DateTimeFormat(locale, { hour: "numeric" }).format(0).match(/AM/);
+  if (typeof window !== "undefined" && navigator) locale = window.navigator?.language;
+
+  if (new Intl.DateTimeFormat(locale, { hour: "numeric" }).format(0).match(/M/)) {
+    localStorage.setItem("timeOption.is24hClock", "false");
+    return false;
+  } else {
+    localStorage.setItem("timeOption.is24hClock", "true");
+    return true;
+  }
```
Comment (joeauyeung) [inline apps/web/lib/timeFormat.ts]: Write to localStorage based on browser locale

File: apps/web/lib/timeFormat.ts [hunk 3]
```diff
@@ -4,9 +4,27 @@
...
+  const locale = "en-US";
+  if (typeof window !== "undefined" && navigator) locale = window.navigator?.language;
+
+  if (!new Intl.DateTimeFormat(locale, { hour: "numeric" }).format(0).match(/M/)) {
+    localStorage.setItem("timeOption.is24hClock", "false");
+    return false;
+  } else {
+    localStorage.setItem("timeOption.is24hClock", "true");
+    return true;
+  }
```
Comment (joeauyeung) [inline apps/web/lib/timeFormat.ts:27]: If not already in localStorage set the value

File: apps/web/pages/getting-started.tsx [hunk 4]
```diff
@@ -214,8 +215,12 @@ export default function Onboarding(props: inferSSRProps<typeof getServerSideProp
         );
       }
     }
+    // Write default timeformat to localStorage
+    const browserTimeFormat = isBrowserLocale24h() ? 24 : 12;
+
```
Comment (joeauyeung) [inline apps/web/pages/getting-started.tsx:220]: When completing onboarding grab the browser locale and set it to localStorage

File: apps/web/pages/getting-started.tsx [hunk 5]
```diff
@@ -214,8 +215,12 @@ export default function Onboarding(props: inferSSRProps<typeof getServerSideProp
...
     await updateUser({
       completedOnboarding: true,
+      timeFormat: browserTimeFormat,
```
Comment (joeauyeung) [inline apps/web/pages/getting-started.tsx:223]: Also set the browser locale in the db

File: apps/web/pages/settings/profile.tsx [hunk 6]
```diff
@@ -179,6 +179,9 @@ function SettingsView(props: ComponentProps<typeof Settings> & { localeProp: str
     const enteredLanguage = selectedLanguage.value;
     const enteredTimeFormat = selectedTimeFormat.value;
 
+    // Write time format to localStorage if available
+    window.localStorage.setItem("timeOption.is24hClock", selectedTimeFormat.value === 12 ? "false" : "true");
```
Comment (joeauyeung) [inline apps/web/pages/settings/profile.tsx:183]: When changing the time format in the settings page also saves the selection to localStorage

**Review by zomars — APPROVED** [review]
Tested and is working flawlessly. Nice work @joeauyeung

**Pull request merged**